pennylane/fourier/reconstruct.py

```diff
@@ -0,0 +1,295 @@
+    `Vidal and Theis (2018) <https://arxiv.org/abs/1812.06323>`_ or
+    `Wierichs et al. (2021) <https://arxiv.org/abs/2107.12390>`_.
+
+    Args:
+        fun (callable): Function to reconstruct
+        num_frequency (int): Number of integer frequencies in ``fun``
+        fun_at_zero (float): Value of ``fun`` at zero; Providing ``fun_at_zero`` saves one
+            evaluation of ``fun``
+
+    Returns:
+        callable: Reconstructed Fourier series with ``num_frequency`` frequencies,
+        as ``qml.numpy`` based function.
+    """
+    if not abs(int(num_frequency)) == num_frequency:
+        raise ValueError(f"num_frequency must be a non-negative integer, got {num_frequency}")
+
+    a, b = (num_frequency + 0.5) / np.pi, 0.5 / np.pi
+    shifts_pos = qml.math.arange(1, num_frequency + 1) / a
+    shifts_neg = -shifts_pos[::-1]
+    fun_at_zero = fun(qml.math.array(0.0)) if fun_at_zero is None else fun_at_zero
+    evals = list(map(fun, shifts_neg)) + [fun_at_zero] + list(map(fun, shifts_pos))
+    shifts = qml.math.concatenate([shifts_neg, [0.0], shifts_pos])
+
+    def _reconstruction(x):
+        """Univariate reconstruction based on equidistant shifts and Dirichlet kernels."""
+        return qml.math.tensordot(
+            qml.math.sinc(a * (x - shifts)) / qml.math.sinc(b * (x - shifts)),
+            evals,
+            axes=[[0], [0]],
+        )
+
+    return _reconstruction
+
+
+_warn_text_fun_at_zero_ignored = (
+    "The provided value of the function at zero will be ignored due to the "
+    "provided shift values. This may lead to additional evaluations of the "
+    "function to be reconstructed."
+)
+
+
+def _reconstruct_gen(fun, spectrum, shifts=None, fun_at_zero=None):
+    r"""Reconstruct a univariate (real-valued) Fourier series with given spectrum.
+    Args:
+        fun (callable): Fourier series to reconstruct with signature ``float -> float``
+        spectrum (Sequence): Frequency spectrum of the Fourier series; non-positive
+            frequencies are ignored
+        shifts (list): Shift angles at which to evaluate ``fun`` for the reconstruction
+            Chosen equidistantly within the interval :math:`[0, 2\pi/f_\text{min}]` if ``shifts=None``
+            where :math:`f_\text{min}` is the smallest frequency in ``spectrum``.
+        fun_at_zero (float): Value of ``fun`` at zero. If :math:`0` is among the ``shifts``
+            and ``fun_at_zero`` is provided, one evaluation of ``fun`` is saved.
+
+    Returns:
+        callable: Reconstructed Fourier series with :math:`R` frequencies in ``spectrum``,
+        as ``qml.numpy`` based function and coinciding with ``fun`` on :math:`2R+1` points.
+    """
+    # pylint: disable=unused-argument
+
+    have_fun_at_zero = fun_at_zero is not None
+    have_shifts = shifts is not None
+    # For an empty/trivial spectrum, the function simply is constant
+    if spectrum in ([], [0.0]):
```

Comment:
```suggestion
    if qml.math.toarray(spectrum).tolist() in ([], [0.0]):
```
this will catch edge cases where `spectrum` is a sequence, but not necessarily a list!

Reply:
Very nice catch! I removed this, though, because the catching of empty/length-1 spectra happens in the outer function `reconstruct` and this additional catching here would only ever be used... in the tests I wrote :D